### ISSUE TITLE
ant design date input type does not work

### ISSUE THREAD

**Issue by alissaVrk** [cypress-io/cypress#16061]
### Current behavior
it seems to be working for a react date input but not for an antd date input
`cy.get("#justDate").type("2020-10-01")` does not change anything

### Desired behavior
for it to change the value

### Test code to reproduce
here is a [repo ](https://github.com/alissaVrk/my-test-app)with reproduction

### Versions

last known version it worked 6.3.0
not working in 7.1.0

**Comment by alissaVrk** [comment]
actually after further investigation.. the problem is not the cypress version but the react version..
with the new cypress version it works with react 6.14.0
but doesn't work with react 17.0.2

it does work in real life and not test

**Comment by jennifer-shehane** [comment]
I can recreate the behavior as described with Cypress not typing the values in the antd inputs in React 17.0.2, but working in React 16.14.0. Maybe the events we trigger are not what is being expected anymore in antd?

**Comment by alissaVrk** [comment]
any update on this? cause for now, I can't upgrade react..

**Comment by jennifer-shehane** [comment]
@alissaVrk No, we likely will not get to fixing this anytime soon. We would welcome any PRs to fix it though.

**Comment by fastfedora** [comment]
I was experiencing the same behavior intermittently with a Material UI date input control. It _feels_ like the `HTMLInputElement` prototype may be being polluted, but I couldn't find where or why. 

Below is a workaround I created that uses a `type` command override and some additional information I gathered during debugging that may be helpful for others continuing to debug this problem.

## Workaround

My workaround was to override the `type` command and use [this React-specific hack](https://github.com/facebook/react/issues/10135#issuecomment-500929024) to set the input value when the input element was one of the elements that Cypress doesn't type into, but sets the value all at once.

This completely fixed the intermittently failing tests that I was experiencing.

```
// The input types where the hack should apply. These are input elements that throw when a partial
// value is set via `input.value`, so needs the entire value to be set at once.
//
// From https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/dom/elements.ts#L41
const inputTypeNeedSingleValueChangeRe = /^(date|time|week|month|datetime-local)$/;

const shouldApplyTypeHack = (element: HTMLElement): boolean => (
  element.tagName?.toLowerCase() === 'input'
    ? inputTypeNeedSingleValueChangeRe.test(element.getAttribute('type')?.toLocaleLowerCase())
    : false
);

type TypeCommand = (
  subject: JQuery<HTMLBodyElement>,
  value: string,
  options: unknown,
) => Cypress.Chainable<JQuery<HTMLBodyElement>>;

function typeCommand(
  originalCommand: TypeCommand,
  subject: JQuery<HTMLBodyElement>,
  value: string,
  options: unknown,
): Cypress.Chainable<JQuery<HTMLBodyElement>> {
  if (subject.length > 1) {
    throw new Error(
      `'type' was called with {{subject.length}} elements, but only a single element is supported`
    );
  }

  const input = subject.get(0);

  if (!shouldApplyTypeHack(input)) {
    return originalCommand(subject, value, options);
  }

  // Hack copied from https://github.com/facebook/react/issues/10135#issuecomment-500929024
  const previousValue = input.value;

  input.value = value;

  if (input._valueTracker) {
    input._valueTracker.setValue(previousValue);
  }

  // 'change' instead of 'input', see
  // https://github.com/facebook/react/issues/11488#issuecomment-381590324
  input.dispatchEvent(new Event('change', { bubbles: true }));

  return subject;
}

Cypress.Commands.overwrite('type', typeCommand);
```

## Debugging Info

### My Test Setup

The problem was occurring with a test suite with 7 tests, 3 of which used `type` within a date input field (tests 2, 3 & 5). The failures only happened intermittently, and the test that failed was not always the same. If I added `only` to a test to isolate the test, it never failed, so it appeared to be some interaction between the tests.

I was able to get a mostly repeatable case where about 90% of time when I ran the entire test suite for the first time, the first two tests would pass, but when I ran them the second time, my 2nd test would fail (the first test to use a date input). In case it's relevant, I was stopping the test after the date input was successfully filled on the 2nd test (I was using `wait` commands to slow down parts of the test), so wasn't allowing the full 7 tests in the test suite to run. 

### My Debugging Setup

For the debugging below, I was editing:

- For React, `react-dom.development.js` in `node_modules/react-dom/cjs`
- For Cypress, `cypress-runner.js` in `~/Library/Caches/Cypress/7.7.0/Cypress.app/Contents/Resources/app/packages/runner/dist` (instructions [here](https://docs.cypress.io/guides/references/troubleshooting#Edit-the-installed-Cypress-code) on finding this location on other people's machines)

To reload the Cypress code, all that was needed was to stop & restart the test within Cypress Studio. To reload the React code, I had to stop and restart Cypress Studio (along with recompiling my app).

### Relevant Codepoints

During debugging, I identified two relevant code points:

- In `react-dom`, `trackValueOnNode` ([source code](https://github.com/facebook/react/blob/main/packages/react-dom/src/client/inputValueTracking.js#L51)) decorates the date input with a new getter & setter
- In `cypress`, `setNativeProp` ([source code](https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/dom/elements.ts#L313)) uses the `value` setter from the `HTMLInputElement` prototype to set the value of the date input, which is being called by [this line](https://github.com/cypress-io/cypress/blob/ce5abe48c0593a3db948773ea23334e7fe581db4/packages/driver/src/cy/keyboard.ts#L809) in the `type` method of the `Keyboard` class

Note that due to the indirection how the functions are called, it's [this line](https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/dom/elements.ts#L325) in `elements.js` that calls the `value` setter for the `HTMLInputElement`. So in the descriptions below, this is the line that I'm debugging with.

### What I Saw

On the 1st successful test run, calling `setNativeProp` sets the value immediately on the date input **without** calling the decorator setter added by `trackValueOnNode` (because `setNativeProp` uses the `HTMLInputElement` prototype's setter and not the setter of the actual node instance).

When I added debugging code directly before and after line 325 in `elements.js`, the result looked like:

```
Calling setNativeProp with '2020-03-01'
Called setNativeProp. Afterwards value of 'node.value' is '2020-03-01'
```

This makes sense, since this should be a "pure" call that only executes the `value` setter for the `HTMLInputElement` prototype. 

On the 2nd test run, where the test fails, calling `setNativeProp` does not set the value immediately. Instead, it somehow calls the setter added by `trackValueOnNode` for the node instance. However, instead of passing the value passed into the setter, it passes an empty string, which causes the field to remain unset.

This is how the same output for the 2nd run would look like:

```
Calling setNativeProp with '2020-03-01'
Calling trackSetValueOnNode with ''
Called setNativeProp. Afterwards value of 'node.value' is ''
```

So somehow on the second run, the `value` setter on the `HTMLInputElement` prototype is calling the setter on the node instance itself (probably via something like `this.value = ''`). 

To try to see if the `HTMLInputElement` prototype was being altered, I annotated the function instance of the setter for the `HTMLInputElement` prototype by calling `annotateDescriptor` at the top of `setNativeProp`:

```
let functionId = 0;

function annotateDescriptor() {
  const descriptor = Object.getOwnPropertyDescriptor(window['HTMLInputElement'].prototype, 'value');

  if (descriptor.set && !descriptor.set.annotated) {
    descriptor.set.functionId = ++functionId;
    descriptor.set.annotated = true;
  }
}
```

The IDs remained the same from the point of annotation through the calls above, so it didn't appear that the prototype was being modified. But I might have missed something in how I did this. 

Note that I also used the same function a) within my test suite and b) within `react-dom` with debugging code to identify where things were being annotated. It appeared as if the `HTMLInputElement` prototype instance used in these three locations (cypress running, React DOM and my tests suite) were 3 different instances. This isn't entirely surprising, since at least the test runner and the application itself are running in different frames (and hence different `window` instances). 

For the test runner, the instance was *not* unloaded between runs (e.g. `descriptor.set.annotated` was already set for the very first call to `annotateDescriptor` on the 2nd test suite run), which may be an additional pointer to the issue. 

The final thing to note is that there was some timing issue also going on. When I added a `debugger` statement just above *or* below line 325 to set into the call to the `value` setter, the problem would never occur. Even if I never tried to set through the code and merely resumed it at that point. Somehow the extra delay of me pausing execution caused the problem to disappear (or that's what it seemed like what happening).

In case it's relevant, I am running my tests on Chromium version 88.0.4324.150 (Official Build) (x86_64) on MacOS 10.15.7.
